File: src/paloma_data/adapters/yelp.py
from __future__ import annotations

from dataclasses import dataclass
import re
import time
from typing import Any, Literal

import httpx

from paloma_data.normalizers import (
    haversine_meters,
    normalize_address,
    normalize_name,
    normalize_phone,
    similarity,
)


YELP_API_BASE = "https://api.yelp.com/v3"
YELP_MATCH_METHOD = "api_business_search_verified_v2"
YELP_RESPONSE_LIMIT_BYTES = 512 * 1_024
YELP_BUSINESS_ID = re.compile(r"^[A-Za-z0-9_-]{1,255}$")

_ALCOHOL_CATEGORY_ALIASES = frozenset(
    {
        "bars",
        "beerbar",
        "beergardens",
        "breweries",
        "brewpubs",
        "cocktailbars",
        "distilleries",
        "divebars",
        "gaybars",
        "hookah_bars",
        "lounges",
        "pubs",
        "speakeasies",
        "sportsbars",
        "tastingrooms",
        "whiskeybars",
        "wine_bars",
        "wineries",
    }
)
_ALCOHOL_CATEGORY_TERMS = (
    "bar",
    "beer",
    "brewery",
    "brewpub",
    "cocktail",
    "distillery",
    "lounge",
    "pub",
    "speakeasy",
    "tasting room",
    "whiskey",
    "wine",
    "winery",
)
_IGNORED_NAME_TOKENS = frozenset({"the", "and", "at", "sf"})


@dataclass(frozen=True, slots=True)
class YelpMatchInput:
    establishment_id: str
    name: str
    address: str
    city: str
    region: str | None
    postal_code: str | None
    country_code: str
    latitude: float
    longitude: float
    phone_e164: str | None


@dataclass(frozen=True, slots=True)
class YelpMatchSelection:
    outcome: Literal["matched", "not_found", "ambiguous", "rejected"]
    reason: str
    provider_place_id: str | None = None
    confidence: float | None = None


@dataclass(frozen=True, slots=True)
class YelpDetailsAudit:
    """Non-persistent projection of a Yelp business-details response."""

    identity_compatible: bool
    identity_reason: str
    currently_operating: bool
    has_phone: bool
    has_hours: bool
    has_price: bool


class YelpAPIError(RuntimeError):
    """A bounded error classification that never includes Yelp response content."""

    def __init__(self, code: str) -> None:
        self.code = code
        super().__init__(f"Yelp API error: {code}")


class YelpPlacesAPI:
    """Resolve durable Yelp IDs without retaining search response attributes.

    Yelp permits business IDs to be retained indefinitely. Search payloads are
    deliberately kept in memory only; rich attributes remain the responsibility
    of the policy-bounded live-details cache.
    """

    def __init__(
        self,
        api_key: str,
        *,
        base_url: str = YELP_API_BASE,
        client: httpx.Client | None = None,
        max_attempts: int = 2,
        sleeper=time.sleep,
    ) -> None:
        if not api_key:
            raise ValueError("YELP_API_KEY is required")
        if max_attempts < 1:
            raise ValueError("max_attempts must be positive")
        self.max_attempts = max_attempts
        self._sleep = sleeper
        self._owns_client = client is None
        self.client = client or httpx.Client(
            base_url=base_url,
            timeout=httpx.Timeout(6.0, connect=3.0),
            headers={
                "Authorization": f"Bearer {api_key}",
                "Accept": "application/json",
                "User-Agent": "paloma-data/0.4",
            },
        )

    def close(self) -> None:
        if self._owns_client:
            self.client.close()

    def __enter__(self) -> "YelpPlacesAPI":
        return self

    def __exit__(self, *_: object) -> None:
        self.close()

    def match(self, expected: YelpMatchInput) -> YelpMatchSelection:
        response = self._get(
            "/businesses/search",
            params={
                "term": expected.name.strip(),
                "latitude": expected.latitude,
                "longitude": expected.longitude,
                "radius": 500,
                "limit": 5,
            },
        )
        _raise_for_status(response)
        if len(response.content) > YELP_RESPONSE_LIMIT_BYTES:
            raise YelpAPIError("invalid_payload")
        try:
            payload = response.json()
        except ValueError as exc:
            raise YelpAPIError("invalid_payload") from exc
        if not isinstance(payload, dict):
            raise YelpAPIError("invalid_payload")
        return select_yelp_business_match(payload, expected)

    def audit_details(
        self,
        provider_place_id: str,
        expected: YelpMatchInput,
    ) -> YelpDetailsAudit:
        """Audit live identity/status/coverage without returning provider attributes."""
        if not YELP_BUSINESS_ID.fullmatch(provider_place_id):
            raise YelpAPIError("invalid_request")
        response = self._get(f"/businesses/{provider_place_id}", params={})
        _raise_for_status(response)
        if len(response.content) > YELP_RESPONSE_LIMIT_BYTES:
            raise YelpAPIError("invalid_payload")
        try:
            payload = response.json()
        except ValueError as exc:
            raise YelpAPIError("invalid_payload") from exc
        if not isinstance(payload, dict):
            raise YelpAPIError("invalid_payload")
        accepted, reason, _, _ = _validate_candidate(payload, expected)
        return YelpDetailsAudit(
            identity_compatible=accepted,
            identity_reason=reason,
            currently_operating=payload.get("is_closed") is False,
            has_phone=bool(_text(payload.get("phone") or payload.get("display_phone"))),
            has_hours=isinstance(payload.get("hours"), list) and bool(payload["hours"]),
            has_price=bool(_text(payload.get("price"))),
        )

    def _get(self, path: str, *, params: dict[str, Any]) -> httpx.Response:
        retryable_statuses = {429, 500, 502, 503, 504}
        for attempt in range(1, self.max_attempts + 1):
            try:
                response = self.client.get(path, params=params)
            except (httpx.TimeoutException, httpx.TransportError) as exc:
                if attempt >= self.max_attempts:
                    raise YelpAPIError("timeout") from exc
                self._sleep(min(2 ** (attempt - 1), 5))
                continue
            if response.status_code not in retryable_statuses or attempt >= self.max_attempts:
                return response
            retry_after = response.headers.get("Retry-After")
            try:
                delay = float(retry_after) if retry_after is not None else 2 ** (attempt - 1)
            except ValueError:
                delay = 2 ** (attempt - 1)
            self._sleep(max(0.0, min(delay, 15.0)))
        raise AssertionError("retry loop did not return")


def select_yelp_business_match(
    payload: dict[str, Any], expected: YelpMatchInput
) -> YelpMatchSelection:
    raw_businesses = payload.get("businesses")
    businesses = raw_businesses if isinstance(raw_businesses, list) else []
    if not businesses:
        return YelpMatchSelection("not_found", "not_found")

    valid: list[tuple[dict[str, Any], float, float]] = []
    rejected_reasons: list[str] = []
    for value in businesses:
        if not isinstance(value, dict):
            rejected_reasons.append("invalid_candidate")
            continue
        accepted, reason, distance, confidence = _validate_candidate(value, expected)
        if accepted:
            valid.append((value, distance, confidence))
        else:
            rejected_reasons.append(reason)

    if len(valid) > 1:
        return YelpMatchSelection("ambiguous", "ambiguous_multiple_candidates")
    if not valid:
        reason = _dominant_rejection_reason(rejected_reasons)
        return YelpMatchSelection("rejected", f"rejected_{reason}")

    business, _, confidence = valid[0]
    business_id = _text(business.get("id"))
    if not business_id or not YELP_BUSINESS_ID.fullmatch(business_id):
        return YelpMatchSelection("rejected", "rejected_invalid_identity")
    return YelpMatchSelection(
        "matched",
        "matched",
        provider_place_id=business_id,
        confidence=confidence,
    )


def _validate_candidate(
    candidate: dict[str, Any], expected: YelpMatchInput
) -> tuple[bool, str, float, float]:
    business_id = _text(candidate.get("id"))
    if not business_id or not YELP_BUSINESS_ID.fullmatch(business_id):
        return False, "invalid_identity", float("inf"), 0.0
    if candidate.get("is_closed") is True:
        return False, "closed", float("inf"), 0.0

    coordinates = candidate.get("coordinates")
    coordinates = coordinates if isinstance(coordinates, dict) else {}
    latitude = _finite_float(coordinates.get("latitude"))
    longitude = _finite_float(coordinates.get("longitude"))
    if latitude is None or longitude is None:
        return False, "missing_coordinates", float("inf"), 0.0
    distance = haversine_meters(
        expected.latitude,
        expected.longitude,
        latitude,
        longitude,
    )
    if distance > 100:
        return False, "location_mismatch", distance, 0.0

    provider_name = _text(candidate.get("name"))
    if not provider_name:
        return False, "name_mismatch", distance, 0.0
    name_score = similarity(normalize_name(expected.name), normalize_name(provider_name))
    name_compatible = _names_are_compatible(expected.name, provider_name)

    location = candidate.get("location")
    location = location if isinstance(location, dict) else {}
    provider_address = _text(location.get("address1"))
    address_score = similarity(
        normalize_address(expected.address),
        normalize_address(provider_address),
    )
    provider_phone = normalize_phone(_text(candidate.get("phone")), expected.country_code)
    exact_phone = bool(
        expected.phone_e164
        and provider_phone
        and expected.phone_e164 == provider_phone
    )

    # Exact/near-exact names are sufficient only at a tight physical location.
    # Looser brand-name compatibility requires an independent address or phone
    # signal. This intentionally trades coverage for false-positive resistance.
    strong_name = name_score >= 0.90
    corroborated_name = name_compatible and (address_score >= 0.86 or exact_phone)
    if not ((strong_name and distance <= 75) or corroborated_name):
        return False, "name_mismatch", distance, 0.0
    if not _has_supported_category(candidate.get("categories")):
        return False, "type_mismatch", distance, 0.0

    if exact_phone:
        confidence = 0.995
    elif address_score >= 0.96 and distance <= 50:
        confidence = 0.99
    elif strong_name and distance <= 50:
        confidence = 0.98
    else:
        confidence = 0.97
    return True, "matched", distance, confidence


def _names_are_compatible(left: str, right: str) -> bool:
    left_tokens = _name_tokens(left)
    right_tokens = _name_tokens(right)
    if not left_tokens or not right_tokens:
        return False
    left_joined = " ".join(left_tokens)
    right_joined = " ".join(right_tokens)
    if (
        left_joined == right_joined
        or left_joined in right_joined
        or right_joined in left_joined
    ):
        return True
    overlap = len(set(left_tokens) & set(right_tokens))
    return overlap / min(len(left_tokens), len(right_tokens)) >= 0.5


def _name_tokens(value: str) -> list[str]:
    return [
        token
        for token in normalize_name(value).split()
        if token and token not in _IGNORED_NAME_TOKENS
    ]


def _has_supported_category(value: Any) -> bool:
    if not isinstance(value, list) or not value:
        return False
    aliases: list[str] = []
    titles: list[str] = []
    for item in value:
        if not isinstance(item, dict):
            continue
        alias = _text(item.get("alias"))
        title = _text(item.get("title"))
        if alias:
            aliases.append(alias.casefold())
        if title:
            titles.append(title.casefold())
    if any(alias in _ALCOHOL_CATEGORY_ALIASES for alias in aliases):
        return True
    return any(term in title for title in titles for term in _ALCOHOL_CATEGORY_TERMS)


def _dominant_rejection_reason(reasons: list[str]) -> str:
    priority = (
        "ambiguous",
        "name_mismatch",
        "location_mismatch",
        "type_mismatch",
        "closed",
        "invalid_identity",
        "missing_coordinates",
        "invalid_candidate",
    )
    return next((reason for reason in priority if reason in reasons), "unusable_candidate")


def _raise_for_status(response: httpx.Response) -> None:
    status = response.status_code
    if 200 <= status < 300:
        return
    if status == 400:
        raise YelpAPIError("invalid_request")
    if status == 401:
        raise YelpAPIError("unauthorized")
    if status == 403:
        raise YelpAPIError("forbidden")
    if status == 404:
        raise YelpAPIError("not_found")
    if status == 429:
        raise YelpAPIError("rate_limited")
    raise YelpAPIError("unavailable")


def _text(value: Any) -> str | None:
    if not isinstance(value, str):
        return None
    stripped = value.strip()
    return stripped or None


def _finite_float(value: Any) -> float | None:
    try:
        parsed = float(value)
    except (TypeError, ValueError):
        return None
    return parsed if parsed == parsed and abs(parsed) != float("inf") else None

File: supabase/functions/venue-live-details/yelp.ts
import {
  e164,
  formattedTime,
  haversineMeters,
  type LiveDetails,
  type LiveFieldRequest,
  namesAreCompatible,
  type PlaceValidation,
  safeHttpUrl,
} from "./domain.ts";

const YELP_API_BASE = "https://api.yelp.com/v3";
const YELP_RESPONSE_LIMIT_BYTES = 512 * 1_024;
const YELP_ERROR_RESPONSE_LIMIT_BYTES = 16 * 1_024;
const YELP_TIMEOUT_MS = 3_500;
const YELP_ERROR_CODE_PATTERN = /^[a-z][a-z0-9_]{0,63}$/;
const YELP_REQUEST_FIELDS = [
  "term",
  "name",
  "address1",
  "address2",
  "address3",
  "city",
  "state",
  "country",
  "postal_code",
  "latitude",
  "longitude",
  "phone",
  "yelp_business_id",
  "limit",
  "radius",
  "sort_by",
  "match_threshold",
  "locale",
  "device_platform",
] as const;

const YELP_ALCOHOL_CATEGORY_ALIASES = new Set([
  "bars",
  "beerbar",
  "beergardens",
  "breweries",
  "brewpubs",
  "cocktailbars",
  "distilleries",
  "divebars",
  "gaybars",
  "hookah_bars",
  "lounges",
  "pubs",
  "speakeasies",
  "sportsbars",
  "tastingrooms",
  "whiskeybars",
  "wine_bars",
  "wineries",
]);

const YELP_ALCOHOL_CATEGORY_TERMS = [
  "bar",
  "beer",
  "brewery",
  "brewpub",
  "cocktail",
  "distillery",
  "lounge",
  "pub",
  "speakeasy",
  "tasting room",
  "whiskey",
  "wine",
  "winery",
];

const YELP_DAY_NAMES: Record<number, string> = {
  0: "Monday",
  1: "Tuesday",
  2: "Wednesday",
  3: "Thursday",
  4: "Friday",
  5: "Saturday",
  6: "Sunday",
};

export type YelpMatchInput = Readonly<{
  name: string;
  address: string;
  city: string;
  region: string | null;
  postalCode: string | null;
  countryCode: string;
  latitude: number;
  longitude: number;
  phoneE164: string | null;
}>;

export type ExpectedYelpPlace = Readonly<{
  yelpBusinessId?: string;
  name: string;
  latitude: number;
  longitude: number;
  phoneE164?: string | null;
}>;

export type YelpMatchSelection =
  | {
    ok: true;
    business: Record<string, unknown>;
    confidence: number;
  }
  | { ok: false; reason: "not_found" | "ambiguous" | "rejected" };

export type YelpApiErrorCode =
  | "invalid_request"
  | "unauthorized"
  | "forbidden"
  | "not_found"
  | "rate_limited"
  | "timeout"
  | "unavailable"
  | "invalid_payload";

export class YelpApiError extends Error {
  constructor(
    readonly code: YelpApiErrorCode,
    readonly providerCode: string | null = null,
  ) {
    super(`Yelp API error: ${code}`);
    this.name = "YelpApiError";
  }
}

export async function fetchYelpBusinessCandidates(
  apiKey: string,
  input: YelpMatchInput,
): Promise<Record<string, unknown>> {
  return await yelpJson(yelpBusinessSearchUrl(input), apiKey);
}

export function yelpBusinessSearchUrl(input: YelpMatchInput): URL {
  const name = input.name.trim();
  if (name.length < 1 || name.length > 300) {
    throw new YelpApiError("invalid_request", "invalid_search_input_name");
  }
  if (
    !Number.isFinite(input.latitude) || Math.abs(input.latitude) > 90 ||
    !Number.isFinite(input.longitude) || Math.abs(input.longitude) > 180
  ) {
    throw new YelpApiError(
      "invalid_request",
      "invalid_search_input_coordinates",
    );
  }

  const url = new URL(`${YELP_API_BASE}/businesses/search`);
  url.searchParams.set("term", name);
  url.searchParams.set("latitude", String(input.latitude));
  url.searchParams.set("longitude", String(input.longitude));
  // Search a wider retrieval window than Paloma's acceptance boundary so
  // provider ranking cannot turn small coordinate drift into a false negative.
  // Candidate acceptance remains fail-closed at 100 metres below.
  url.searchParams.set("radius", "500");
  url.searchParams.set("limit", "5");
  return url;
}

export async function fetchYelpBusinessDetails(
  apiKey: string,
  businessId: string,
): Promise<Record<string, unknown>> {
  return await yelpJson(yelpBusinessDetailsUrl(businessId), apiKey);
}

export function yelpBusinessDetailsUrl(businessId: string): URL {
  const url = new URL(
    `${YELP_API_BASE}/businesses/${encodeURIComponent(businessId)}`,
  );
  url.searchParams.set("locale", "en_US");
  return url;
}

export function selectYelpBusinessMatch(
  payload: Record<string, unknown>,
  expected: ExpectedYelpPlace,
): YelpMatchSelection {
  const businesses = Array.isArray(payload.businesses)
    ? payload.businesses
    : [];
  const valid: Array<{
    business: Record<string, unknown>;
    distanceMeters: number;
  }> = [];
  for (const value of businesses) {
    const business = object(value);
    if (!business) continue;
    const validation = validateYelpPlace(business, expected, false);
    if (validation.ok) {
      valid.push({
        business,
        distanceMeters: validation.distanceMeters,
      });
    }
  }
  if (valid.length === 0) {
    return {
      ok: false,
      reason: businesses.length === 0 ? "not_found" : "rejected",
    };
  }
  if (valid.length !== 1) return { ok: false, reason: "ambiguous" };
  return {
    ok: true,
    business: valid[0].business,
    confidence: valid[0].distanceMeters <= 50 ? 0.99 : 0.97,
  };
}

export function validateYelpPlace(
  raw: Record<string, unknown>,
  expected: ExpectedYelpPlace,
  requireAttribution = true,
): PlaceValidation {
  const returnedId = text(raw.id);
  if (!returnedId || returnedId.length > 255) {
    return { ok: false, reason: "missing_identity" };
  }
  if (expected.yelpBusinessId && returnedId !== expected.yelpBusinessId) {
    return { ok: false, reason: "identity_mismatch" };
  }
  if (raw.is_closed === true) return { ok: false, reason: "closed" };

  const coordinates = object(raw.coordinates);
  const latitude = finiteNumber(coordinates?.latitude);
  const longitude = finiteNumber(coordinates?.longitude);
  if (
    latitude === null || longitude === null || Math.abs(latitude) > 90 ||
    Math.abs(longitude) > 180
  ) {
    return { ok: false, reason: "missing_coordinates" };
  }
  const distanceMeters = haversineMeters(
    expected.latitude,
    expected.longitude,
    latitude,
    longitude,
  );
  if (distanceMeters > 100) {
    return { ok: false, reason: "location_mismatch" };
  }

  const providerName = text(raw.name);
  const exactPhone = Boolean(
    expected.phoneE164 && e164(raw.phone) === expected.phoneE164,
  );
  if (
    !providerName ||
    !strongYelpNameMatch(expected.name, providerName) &&
      !(exactPhone && namesAreCompatible(expected.name, providerName))
  ) {
    return { ok: false, reason: "name_mismatch" };
  }
  if (!hasSupportedYelpCategory(raw.categories)) {
    return { ok: false, reason: "type_mismatch" };
  }
  if (requireAttribution && !yelpAttributionUrl(raw)) {
    return { ok: false, reason: "missing_attribution" };
  }
  return { ok: true, distanceMeters };
}

export function projectYelpLiveDetails(
  raw: Record<string, unknown>,
  requested: LiveFieldRequest,
): LiveDetails {
  const price = requested.price ? text(raw.price) : null;
  return {
    // Yelp's primary listing photo is display-only, subject to the provider
    // response policy and the same-screen Yelp credit returned below.
    cover_image_url: yelpCoverImageUrl(raw),
    phone_e164: requested.phone ? e164(raw.phone) : null,
    // Yelp's `url` is the required Yelp attribution link, not the venue's own
    // website. Never relabel it as consumer contact data.
    website_url: null,
    hours: requested.hours ? normalizedYelpHours(raw.hours) : null,
    open_now: requested.hours ? yelpOpenNow(raw.hours) : null,
    price_level: price && /^\${1,4}$/.test(price) ? price.length : null,
    // Base Yelp details do not provide an objective setting taxonomy that maps
    // safely to Paloma's durable setting slugs.
    setting_slugs: [],
  };
}

export function yelpCoverImageUrl(raw: Record<string, unknown>): string | null {
  const supplied = safeHttpUrl(raw.image_url);
  if (!supplied) return null;
  const url = new URL(supplied);
  const host = url.hostname.toLowerCase();
  // Yelp documents CDN-hosted Business Details image URLs. Refusing every
  // other host makes this a Yelp image field, rather than a generic remote URL
  // that could slip into a provider response and be displayed without review.
  if (host !== "yelpcdn.com" && !host.endsWith(".yelpcdn.com")) return null;
  // Yelp documents both HTTP and HTTPS samples. Never hand an HTTP image to
  // the iPhone, where App Transport Security would reject it anyway.
  url.protocol = "https:";
  return url.toString();
}

export function yelpAttributionUrl(
  raw: Record<string, unknown>,
): string | null {
  const supplied = safeHttpUrl(raw.url);
  if (!supplied) return null;
  const host = new URL(supplied).hostname.toLowerCase();
  return host === "yelp.com" || host.endsWith(".yelp.com") ? supplied : null;
}

export function yelpBusinessId(
  raw: Record<string, unknown>,
): string | null {
  const value = text(raw.id);
  return value && value.length <= 255 ? value : null;
}

async function yelpJson(
  url: URL,
  apiKey: string,
): Promise<Record<string, unknown>> {
  const controller = new AbortController();
  const timeout = setTimeout(() => controller.abort(), YELP_TIMEOUT_MS);
  let response: Response;
  try {
    response = await fetch(url, {
      method: "GET",
      headers: {
        "Authorization": `Bearer ${apiKey}`,
        "Accept": "application/json",
      },
      cache: "no-store",
      signal: controller.signal,
    });
  } catch (error) {
    throw new YelpApiError(
      error instanceof Error && error.name === "AbortError"
        ? "timeout"
        : "unavailable",
    );
  } finally {
    clearTimeout(timeout);
  }

  const responseError = yelpApiErrorCodeForStatus(response.status);
  if (responseError) {
    throw new YelpApiError(
      responseError,
      await yelpProviderErrorCodeFromResponse(response),
    );
  }

  const contentLength = Number(response.headers.get("content-length") ?? 0);
  if (
    Number.isFinite(contentLength) && contentLength > YELP_RESPONSE_LIMIT_BYTES
  ) {
    throw new YelpApiError("invalid_payload");
  }
  let body: string;
  try {
    body = await response.text();
  } catch {
    throw new YelpApiError("invalid_payload");
  }
  if (new TextEncoder().encode(body).byteLength > YELP_RESPONSE_LIMIT_BYTES) {
    throw new YelpApiError("invalid_payload");
  }
  try {
    const payload = object(JSON.parse(body));
    if (!payload) throw new YelpApiError("invalid_payload");
    return payload;
  } catch (error) {
    if (error instanceof YelpApiError) throw error;
    throw new YelpApiError("invalid_payload");
  }
}

async function yelpProviderErrorCodeFromResponse(
  response: Response,
): Promise<string | null> {
  const contentLength = Number(response.headers.get("content-length") ?? 0);
  if (
    Number.isFinite(contentLength) &&
    contentLength > YELP_ERROR_RESPONSE_LIMIT_BYTES
  ) return null;
  try {
    const body = await response.text();
    if (
      new TextEncoder().encode(body).byteLength >
        YELP_ERROR_RESPONSE_LIMIT_BYTES
    ) return null;
    return yelpProviderErrorCode(JSON.parse(body));
  } catch {
    return null;
  }
}

export function yelpProviderErrorCode(value: unknown): string | null {
  const payload = object(value);
  const current = object(payload?.error);
  const legacy = Array.isArray(payload?.errors)
    ? object(payload.errors[0])
    : null;
  const rawCode = text(current?.code) ?? text(legacy?.error_code);
  const description = text(current?.description) ??
    text(legacy?.error_message);
  const normalized = rawCode?.toLowerCase().replace(/[^a-z0-9]+/g, "_")
    .replace(/^_+|_+$/g, "") ?? "";
  if (!normalized) return null;

  const lowerDescription = description?.toLowerCase() ?? "";
  const field = YELP_REQUEST_FIELDS.find((candidate) =>
    new RegExp(`(^|[^a-z0-9_])${candidate}([^a-z0-9_]|$)`).test(
      lowerDescription,
    )
  );
  const candidate = `yelp_${normalized}${field ? `_${field}` : ""}`
    .slice(0, 64)
    .replace(/_+$/g, "");
  return YELP_ERROR_CODE_PATTERN.test(candidate) ? candidate : null;
}

export function yelpApiErrorCodeForStatus(
  status: number,
): YelpApiErrorCode | null {
  if (status >= 200 && status < 300) return null;
  if (status === 400) return "invalid_request";
  if (status === 401) return "unauthorized";
  if (status === 403) return "forbidden";
  if (status === 404) return "not_found";
  if (status === 429) return "rate_limited";
  return "unavailable";
}

function normalizedYelpHours(
  value: unknown,
): { weekday_text: string[] } | null {
  if (!Array.isArray(value)) return null;
  const group =
    value.map(object).find((row) => row?.hours_type === "REGULAR") ??
      value.map(object).find(Boolean);
  const periods = Array.isArray(group?.open) ? group.open : [];
  const byDay = new Map<number, string[]>();
  for (const value of periods) {
    const period = object(value);
    const day = finiteNumber(period?.day);
    const open = formattedTime(period?.start);
    const close = formattedTime(period?.end);
    if (
      day === null || !Number.isInteger(day) || !YELP_DAY_NAMES[day] ||
      !open || !close
    ) {
      continue;
    }
    const values = byDay.get(day) ?? [];
    values.push(`${open}–${close}`);
    byDay.set(day, values);
  }
  const weekdayText = [...byDay.entries()]
    .sort(([left], [right]) => left - right)
    .map(([day, periods]) => `${YELP_DAY_NAMES[day]}: ${periods.join(", ")}`);
  return weekdayText.length > 0 ? { weekday_text: weekdayText } : null;
}

function yelpOpenNow(value: unknown): boolean | null {
  if (!Array.isArray(value)) return null;
  const group =
    value.map(object).find((row) => row?.hours_type === "REGULAR") ??
      value.map(object).find(Boolean);
  return typeof group?.is_open_now === "boolean" ? group.is_open_now : null;
}

function hasSupportedYelpCategory(value: unknown): boolean {
  if (!Array.isArray(value) || value.length === 0) return false;
  return value.some((item) => {
    const category = object(item);
    const alias = text(category?.alias)?.toLowerCase() ?? "";
    const title = text(category?.title)?.toLowerCase() ?? "";
    return YELP_ALCOHOL_CATEGORY_ALIASES.has(alias) ||
      YELP_ALCOHOL_CATEGORY_TERMS.some((term) => title.includes(term));
  });
}

function strongYelpNameMatch(left: string, right: string): boolean {
  const leftTokens = yelpNameTokens(left);
  const rightTokens = yelpNameTokens(right);
  if (leftTokens.length === 0 || rightTokens.length === 0) return false;
  const leftJoined = leftTokens.join(" ");
  const rightJoined = rightTokens.join(" ");
  if (
    leftJoined === rightJoined || leftJoined.includes(rightJoined) ||
    rightJoined.includes(leftJoined)
  ) return true;
  const rightSet = new Set(rightTokens);
  const overlap = leftTokens.filter((token) => rightSet.has(token)).length;
  return overlap / Math.min(leftTokens.length, rightTokens.length) >= 0.75;
}

function yelpNameTokens(value: string): string[] {
  return value
    .normalize("NFKD")
    .replace(/[\u0300-\u036f]/g, "")
    .toLowerCase()
    .replace(/&/g, " and ")
    .replace(/[^a-z0-9]+/g, " ")
    .trim()
    .split(/\s+/)
    .filter((token) => token && !["the", "and", "at", "sf"].includes(token));
}

function object(value: unknown): Record<string, unknown> | null {
  return value !== null && typeof value === "object" && !Array.isArray(value)
    ? value as Record<string, unknown>
    : null;
}

function text(value: unknown): string | null {
  if (typeof value !== "string") return null;
  const trimmed = value.trim();
  return trimmed || null;
}

function finiteNumber(value: unknown): number | null {
  const parsed = typeof value === "number" ? value : Number(value);
  return Number.isFinite(parsed) ? parsed : null;
}
